src/main/java/com/conveyal/datatools/manager/jobs/ProcessSingleFeedJob.java

```diff
@@ -48,21 +65,102 @@ public String getFeedSourceId () {
     @Override
     public void jobLogic () {
         LOG.info("Processing feed for {}", feedVersion.id);
+        feedSource = feedVersion.parentFeedSource();
+        // If there are any feed transformations that apply to the VERSION_CLONE retrieval method, we need to add a
+        // stage to jobFinished (perhaps) to clone the feed version (as long as this feed version has not already been
+        // cloned). Once that feed version is processed, the appropriate transformations will apply themselves.
+        shouldClone = !feedVersion.retrievalMethod.equals(VERSION_CLONE) &&
+            feedSource.hasRulesForRetrievalMethod(VERSION_CLONE);
+        FeedTransformRules rules = feedSource.getRulesForRetrievalMethod(feedVersion.retrievalMethod);
+        boolean shouldTransform = rules != null;
+        if (shouldTransform) {
+            // Run zip transformations before load to handle any operations that must be applied directly to the zip file.
+            List<ZipTransformation> zipTransformations = rules.getActiveTransformations(feedVersion, ZipTransformation.class);
+            for (ZipTransformation transformation : zipTransformations) {
+                ArbitraryTransformJob zipTransform = new ArbitraryTransformJob(owner, feedVersion.retrieveGtfsFile(), transformation);
+                // Run transform job in line so we can monitor the error status before load/validate begins.
+                zipTransform.run();
+                // Short circuit the feed load/validate if a pre-load transform fails.
+                if (zipTransform.status.error) return;
+            }
+        }
 
         // First, load the feed into database. During this stage, the GTFS file will be uploaded to S3 (and deleted locally).
         addNextJob(new LoadFeedJob(feedVersion, owner, isNewVersion));
 
         // Next, validate the feed.
         addNextJob(new ValidateFeedJob(feedVersion, owner, isNewVersion));
 
+        // We only need to snapshot the feed if there are transformations at the database level. In the case that there
+        // are, the snapshot namespace will be the target of these modifications. If we were to apply the modifications
+        // to the original feed version's namespace, the namespace and the zip file would become out of sync, which
+        // violates a core expectation for this application: that each zip file represents exactly its database
+        // representation.
```

Comment:
This confuses me somewhat, because there is a possibility to transform a zip file in a previous block of code in this method with a ZipTransformation. Wouldn't this violate the stated expectation that each zip file that is inputted has a database representation?

Reply:
In the case that the zip transformation applies to incoming GTFS. That incoming GTFS is modified before it is ever loaded into the database, so the resulting zip and db match one another. Your comment does make me think of a potential pitfall to this approach: If we're fetching the feed automatically and applying a zip transformation, we may not be able to recognize the next time we fetch GTFS that the version we're loading is the same version we previously loaded (because we have modified it).